akka-remote/src/main/scala/akka/remote/artery/Association.scala

```diff
@@ -958,9 +958,14 @@ private[remote] class Association(
 
     implicit val ec = materializer.executionContext
     streamCompleted.foreach { _ =>
-      // shutdown as expected
-      // countDown the latch in case threads are waiting on the latch in outboundControlIngress method
-      materializing.countDown()
+      if (transport.isShutdown || isRemovedAfterQuarantined()) {
+        // shutdown as expected
+        // countDown the latch in case threads are waiting on the latch in outboundControlIngress method
+        materializing.countDown()
+      } else {
+        log.debug("{} to [{}] was completed. It will be restarted if used again.", streamName, remoteAddress)
+        lazyRestart()
+      }
```

Comment:
I don't know how to reproduce this scenario for real but simulated it with a `take(20)` in the connection flow and verified that sending more than 20 messages triggered this case and messages afterwards would also be delivered, which wasn't the case before the fix.